pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -822,6 +845,53 @@ func (pl *DynamicResources) PreBind(ctx context.Context, cs fwk.CycleState, pod
 			state.claims[index] = claim
 		}
 	}
+
+	if !pl.enableDeviceBindingConditions || !pl.enableDeviceStatus {
+		// If we get here, we know that reserving the claim for
+		// the pod worked and we can proceed with binding it.
+		return nil
+	}
+
+	// We need to check if the device is attached to the node.
+	needWait := hasBindingConditions(state)
+
+	// If no device needs to be prepared, we can return early.
+	if !needWait {
+		return nil
+	}
+
+	// We need to decide how long we should wait for the device to be attached to the node.
+	const timeoutDefaultSeconds int64 = 600
+	timeoutMax := 20 * time.Minute
+	timeout := 0 * time.Second
+	timeStartWaiting := time.Now()
+
+	for _, claim := range state.claims {
+		for _, device := range claim.Status.Allocation.Devices.Results {
+			if deviceTimeout := time.Duration(ptr.Deref(device.BindingTimeoutSeconds, timeoutDefaultSeconds)) * time.Second; timeout < deviceTimeout {
+				timeout = deviceTimeout
+			}
+		}
+	}
+	if timeout > timeoutMax {
+		timeout = timeoutMax
+	}
+	deadlineCtx, cancelFunc := context.WithDeadline(ctx, timeStartWaiting.Add(timeout))
+	defer cancelFunc()
```

Comment:
PollUntilContextTimeout below already applies the timeout. There's no need to also to do it via the context.

Reply:
I have updated.